### PR TITLE
897 new add initialization method from flows for lct model 1

### LINKED ISSUE #897: Add initialization method from flows for LCT model
### Feature description

Initialization method with flows for lct model, initialization method from IDE model. Can be used to generate consistent initial numbers for LCT and IDE model.

### Additional context

Code can be copied from #680 
Files:
- cpp/models/lct_secir/initialization.cpp 
- cpp/tests/test_lct_initializer.cpp
- add example?

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project.

### PR DESCRIPTION
**Pull request by lenaploetzke**
# Changes and Information

Please briefly list the changes (main added features, changed items, or corrected bugs) made:

  -  added functionality to compute an initial vector for the LCT-SECIR model based on flows

If need be, add additional information and what the reviewer should look out for in particular:

-/-

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [ ] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

Closes #897.

### CODE REVIEW

File: cpp/simulations/munich_graph_sim.cpp [hunk 1]
```diff
@@ -287,4 +287,4 @@ int main(int argc, char** argv)
         return -1;
     }
     return 0;
-}
+}
```
Comment (lenaploetzke) [inline cpp/simulations/munich_graph_sim.cpp:290]: Nothing happend here, i dont know why it is marked

File: cpp/examples/lct_secir.cpp [hunk 2]
```diff
@@ -126,6 +69,109 @@ int main()
...
+        vec_flows[(int)mio::lsecir::InfectionTransition::InfectedSymptomsToRecovered]          = 4.0;
+        vec_flows[(int)mio::lsecir::InfectionTransition::InfectedSevereToInfectedCritical]     = 1.0;
+        vec_flows[(int)mio::lsecir::InfectionTransition::InfectedSevereToRecovered]            = 1.0;
+        vec_flows[(int)mio::lsecir::InfectionTransition::InfectedCriticalToDead]               = 1.0;
+        vec_flows[(int)mio::lsecir::InfectionTransition::InfectedCriticalToRecovered]          = 1.0;
+        // Add initial time point to time series.
+        flows.add_time_point(-110, vec_flows);
+        // Add further time points until time 0.
+        while (flows.get_last_time() < -1e-10) {
+            flows.add_time_point(flows.get_last_time() + dt, vec_flows);
```
Comment (annawendler) [inline cpp/examples/lct_secir.cpp:95]: I think here you should make vec_flows dependent on dt as in the IDE example. For example when I set `dt=0.001` this leads to an error.

File: cpp/examples/lct_secir.cpp [hunk 3]
```diff
@@ -126,6 +69,109 @@ int main()
...
+        }
+
+        // Set initialization vector for the LCT model.
+        mio::lsecir::Initializer<Model> initializer(std::move(flows), model);
+        initializer.set_tol_for_support_max(1e-6);
+        auto status = initializer.compute_initialization_vector(1000000, 10, 16000);
```
Comment (annawendler) [inline cpp/examples/lct_secir.cpp:101]: I would find this example easier to read if you would declare total_population, deaths and total_confirmed_cases separately and then use these variables as arguments. But of course both ways is fine :)

File: cpp/examples/lct_secir.cpp [hunk 4]
```diff
@@ -126,6 +69,109 @@ int main()
...
+            (initial_populations[(int)LctState::InfectionState::InfectedSevere].size() !=
+             (size_t)LctState::get_num_subcompartments<LctState::InfectionState::InfectedSevere>()) ||
+            (initial_populations[(int)LctState::InfectionState::InfectedCritical].size() !=
+             (size_t)LctState::get_num_subcompartments<LctState::InfectionState::InfectedCritical>()) ||
+            (initial_populations[(int)LctState::InfectionState::Recovered].size() !=
+             (size_t)LctState::get_num_subcompartments<LctState::InfectionState::Recovered>()) ||
+            (initial_populations[(int)LctState::InfectionState::Dead].size() !=
+             (size_t)LctState::get_num_subcompartments<LctState::InfectionState::Dead>())) {
+            mio::log_error(
+                "The length of at least one vector in initial_populations does not match the related number of "
```
Comment (annawendler) [inline cpp/examples/lct_secir.cpp:138]:
```suggestion
                "The length of at least one vector in initial_populations does not match the corresponding number of "
```

File: cpp/models/lct_secir/initializer_flows.h [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+#include "memilio/utils/logging.h"
+#include "memilio/epidemiology/state_age_function.h"
+
+namespace mio
+{
+namespace lsecir
+{
+
+/**
+ * @brief Class that can be used to compute an initialization vector out of flows for a LCT Model.
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:40]: I think it makes sense to explain the idea more in detail. For example in the ODE initialization only the flow from C to I is used and here you are using all possible flows. Also, you could mention that you need a certain number of time points that depends on the relevant are of the Erlang density function.

File: cpp/models/lct_secir/initializer_flows.h [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+        //S
+        init[0] = total_population - init.segment(1, LctState::Count - 2).sum() - deaths;
+
+        //D
+        init[LctState::Count - 1] = deaths;
+
+        for (int i = 0; i < LctState::Count; i++) {
+            if (init[i] < 0) {
+                log_error("Initialization failed. Not enough time points for the transitions are given. More than {} "
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:128]: Is it possible that the size of the subcompartments negative although there were enough time points, for example if there are negative flows given?

File: cpp/models/lct_secir/initializer_flows.h [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+                          "are needed but "
+                          "just {} are given.",
+                          -init[i], m_flows.get_num_time_points());
+                return true;
+            }
+        }
+
+        // Update initial value vector of the mpdel.
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:136]:
```suggestion
        // Update initial value vector of the model.
```

File: cpp/models/lct_secir/initializer_flows.h [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        m_model.set_initial_values(init);
+
+        // Check if constraints were fulfilled.
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:139]:
```suggestion
        // Check if constraints are fulfilled.
```

File: cpp/models/lct_secir/initializer_flows.h [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+     * @param[in] new_tol New tolerance.
+     */
+    void set_tol_for_support_max(ScalarType new_tol)
+    {
+        m_tol = new_tol;
+    }
+
+private:
+    /**
+     * @brief Checks constraints of the Initializer inclusive check for the model.
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:155]:
```suggestion
     * @brief Checks constraints of the Initializer including checks for the model.
```

File: cpp/models/lct_secir/initializer_flows.h [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+                return true;
+            }
+        }
+
+        if (!(m_dt < 1)) {
+            log_warning("Step size was set very large. The result could be distorted.");
+            return true;
+        }
+
+        // Check if model is valid and the calculated initial value vector was valid.
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:182]:
```suggestion
        // Check if model is valid and the calculated initial value vector is valid.
```

File: cpp/models/lct_secir/initializer_flows.h [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+        return m_model.check_constraints();
+    }
+
+    /**
+     * @brief Computes a vector with the number of people in each subcompartment for one InfectionState.
+     *
+     * With this function, partial result of compute_initialization_vector are achieved.
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:189]:
```suggestion
     * This function is used in compute_initialization_vector.
```

File: cpp/models/lct_secir/initializer_flows.h [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+     * @param[in] idx_incoming_flow Index of the flow which is relevant for the calculation, so the flow to the InfectionState.
+     * @param[in] transition_rate Specifies the transition rate of the InfectionState. Is equal to 1 / (expected Time in the InfectionState).
+     * @tparam State The InfectionState for which the partial result should be calculated.
+     * @return Vector with a possible initialization for the subcompartments of the InfectionState State. 
+     *      Subcompartment is set to -1 if calculation was not possible.
+     */
+    template <InfectionState State>
+    Eigen::VectorXd compute_compartment(Eigen::Index idx_incoming_flow, ScalarType transition_rate) const
+    {
+        int num_infectionstates = LctState::template get_num_subcompartments<State>();
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:200]:
```suggestion
        int num_subcompartments = LctState::template get_num_subcompartments<State>();
```
I find num_infectionstates confusing here.

File: cpp/models/lct_secir/initializer_flows.h [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+        ErlangDensity erlang(1, 1. / (num_infectionstates * transition_rate));
+
+        // Initialize other relevant parameters.
+        ScalarType calc_time{0};
+        Eigen::Index calc_time_index{0};
+        ScalarType state_age{0};
+        ScalarType sum{0};
+        Eigen::Index num_time_points = m_flows.get_num_time_points();
+
+        // Calculate number of people in each subcomaprtment.
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:213]:
```suggestion
        // Calculate number of people in each subcompartment.
```

File: cpp/models/lct_secir/initializer_flows.h [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+
+            // Determine relevant calculation area and corresponding index.
+            calc_time       = erlang.get_support_max(m_dt, m_tol);
+            calc_time_index = (Eigen::Index)std::ceil(calc_time / m_dt) - 1;
+
+            if (num_time_points < calc_time_index) {
+                // Initialization failed. Not enough time points for the transitions are given.
+                // Store the number of necessary time points in the vector and throw one error at the compute_initialization_vector() function,
+                // so that the error is not displayed several times.
+                subcompartments[j] = -(double)calc_time_index;
```
Comment (annawendler) [inline cpp/models/lct_secir/initializer_flows.h:226]:
```suggestion
                subcompartments[j] = -(ScalarType)calc_time_index;
```

File: cpp/tests/test_lct_initializer_flows.cpp [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    vec_init[(int)mio::lsecir::InfectionTransition::InfectedSymptomsToRecovered]          = 4.0;
+    // Add initial time point to time series.
+    init.add_time_point(-130, vec_init);
+    // Add further time points until time 0.
+    while (init.get_last_time() < 0) {
+        vec_init *= 1.01;
+        init.add_time_point(init.get_last_time() + dt, vec_init);
+    }
+
+    // Calculate initial vector and compare with previous reult.
```
Comment (annawendler) [inline cpp/tests/test_lct_initializer_flows.cpp:88]:
```suggestion
    // Calculate initial vector and compare with previous result.
```

**Review by annawendler — CHANGES_REQUESTED** [review]
Looks really nice :)

I was thinking that initializer_subcompartments might be a more suitable name than initializer_flows as you are calculating initial values of the subcompartments for given flows.

**Comment by lenaploetzke** [comment]
> I was thinking that initializer_subcompartments might be a more suitable name than initializer_flows as you are calculating initial values of the subcompartments for given flows.

I do not think that initializer_subcompartments is a better name, because e.g. the realistic initialization also initializes subcompartments. The naming should include that we are using flows for initialization.

File: cpp/examples/lct_secir.cpp (hunk 2 above)
Comment (lenaploetzke) [reply cpp/examples/lct_secir.cpp:95]: Yes, thanks a lot!

File: cpp/models/lct_secir/initializer_flows.h (hunk 6 above)
Comment (lenaploetzke) [reply cpp/models/lct_secir/initializer_flows.h:128]: Yes it is, thanks, I adapted this. I think it is okay now.

File: cpp/models/lct_secir/initializer_flows.h (hunk 5 above)
Comment (lenaploetzke) [reply cpp/models/lct_secir/initializer_flows.h:40]: Are you okay with this version?

**Comment by lenaploetzke** [comment]
The LCT model has been updated in the meantime, so small changes had to be made here

**Review by annawendler — APPROVED** [review]

**Pull request merged**